File: backend/app/services/gemini.py
import asyncio
import json
import time

import google.generativeai as genai
from tenacity import retry, stop_after_attempt, wait_exponential

from app.config import settings

_genai_configured = False
_failure_count = 0
_first_failure_time = 0.0
_circuit_open_until = 0.0

ALLOWED_PRIORITIES = {"high", "medium", "low"}
ALLOWED_RISKS = {"high", "medium", "low"}
ALLOWED_CATEGORIES = {"primary", "updates", "promotions", "social", "forums"}


def _ensure_configured():
    global _genai_configured
    if not _genai_configured:
        genai.configure(api_key=settings.gemini_api_key)
        _genai_configured = True


def _record_failure():
    global _failure_count, _first_failure_time, _circuit_open_until
    now = time.time()
    if _first_failure_time == 0 or now - _first_failure_time > 60:
        _first_failure_time = now
        _failure_count = 0
    _failure_count += 1
    if _failure_count >= 5:
        _circuit_open_until = now + 120
        _failure_count = 0
        _first_failure_time = 0


def _circuit_open() -> bool:
    return time.time() < _circuit_open_until


def _bounded_text(value: str, limit: int = 6000) -> str:
    cleaned = " ".join((value or "").split())
    return cleaned[:limit]


def _extract_json(text: str) -> dict | list:
    candidate = (text or "").strip()
    if candidate.startswith("```"):
        candidate = candidate.strip("`")
        if "\n" in candidate:
            candidate = candidate.split("\n", 1)[1]
        if candidate.endswith("```"):
            candidate = candidate[:-3]
    try:
        return json.loads(candidate)
    except json.JSONDecodeError:
        start_obj = candidate.find("{")
        end_obj = candidate.rfind("}")
        start_list = candidate.find("[")
        end_list = candidate.rfind("]")
        if start_obj != -1 and end_obj != -1 and end_obj > start_obj:
            return json.loads(candidate[start_obj : end_obj + 1])
        if start_list != -1 and end_list != -1 and end_list > start_list:
            return json.loads(candidate[start_list : end_list + 1])
        raise


@retry(stop=stop_after_attempt(2), wait=wait_exponential(multiplier=1, min=1, max=2))
async def _call_json(model_name: str, prompt: str) -> dict | list:
    if _circuit_open():
        raise RuntimeError("Gemini circuit open")
    _ensure_configured()
    model = genai.GenerativeModel(model_name)

    def _run():
        response = model.generate_content(
            prompt,
            generation_config={"response_mime_type": "application/json", "temperature": 0.15},
        )
        return response.text

    try:
        text = await asyncio.wait_for(asyncio.to_thread(_run), timeout=15)
        return _extract_json(text)
    except Exception:
        _record_failure()
        raise


def _flash_model() -> str:
    return settings.gemini_model or "gemini-1.5-flash"


def _normalize_priority(data: dict | None) -> dict | None:
    if not isinstance(data, dict):
        return None
    priority = str(data.get("priority", "")).lower()
    if priority not in ALLOWED_PRIORITIES:
        return None
    reason = str(data.get("reason", "")).strip()[:240] or "Priority inferred from message content."
    confidence_raw = data.get("confidence", 0.5)
    try:
        confidence = max(0.0, min(1.0, float(confidence_raw)))
    except (TypeError, ValueError):
        confidence = 0.5
    return {"priority": priority, "reason": reason, "confidence": confidence}


def _normalize_risk(data: dict | None) -> dict | None:
    if not isinstance(data, dict):
        return None
    risk = str(data.get("risk", "")).lower()
    if risk not in ALLOWED_RISKS:
        return None
    reasons = [str(reason).strip()[:160] for reason in data.get("reasons", []) if str(reason).strip()]
    return {"risk": risk, "reasons": reasons[:5]}


async def summarize_email(text: str) -> list[str] | None:
    bounded = _bounded_text(text)
    if not bounded:
        return None
    prompt = (
        "Summarize this student-facing email into exactly 3 concise bullet points. "
        "Each bullet must be 6 to 15 words. Return JSON only as {\"bullets\": [..]}.\n\n"
        f"EMAIL:\n{bounded}"
    )
    data = await _call_json(_flash_model(), prompt)
    bullets = data.get("bullets") if isinstance(data, dict) else None
    if not isinstance(bullets, list):
        return None
    cleaned = [str(item).strip()[:120] for item in bullets if str(item).strip()]
    if len(cleaned) != 3:
        return None
    return cleaned


async def priority_analysis(sender: str, subject: str, body: str) -> dict | None:
    prompt = (
        "Assign priority for a student inbox. "
        "Consider deadlines, account actions, schedule changes, exams, finances, and urgency. "
        "Return JSON only as {priority: 'high'|'medium'|'low', reason: string, confidence: float}.\n\n"
        f"SENDER: {_bounded_text(sender, 200)}\nSUBJECT: {_bounded_text(subject, 300)}\nBODY: {_bounded_text(body)}"
    )
    return _normalize_priority(await _call_json(_flash_model(), prompt))


async def category_classification(sender: str, subject: str, snippet: str) -> str | None:
    prompt = (
        "Classify a student email into one of: primary, updates, promotions, social, forums. "
        "Return JSON only as {category: 'primary'|'updates'|'promotions'|'social'|'forums'}.\n\n"
        f"SENDER: {_bounded_text(sender, 200)}\nSUBJECT: {_bounded_text(subject, 300)}\nSNIPPET: {_bounded_text(snippet, 600)}"
    )
    data = await _call_json(_flash_model(), prompt)
    if isinstance(data, dict):
        category = str(data.get("category", "")).lower()
        if category in ALLOWED_CATEGORIES:
            return category
    return None


async def phishing_analysis(sender: str, subject: str, body: str) -> dict | None:
    prompt = (
        "Analyze this email for phishing, impersonation, credential theft, payment fraud, or manipulative urgency. "
        "Be conservative and return JSON only as {risk: 'low'|'medium'|'high', reasons: [string]}.\n\n"
        f"SENDER: {_bounded_text(sender, 200)}\nSUBJECT: {_bounded_text(subject, 300)}\nBODY: {_bounded_text(body)}"
    )
    return _normalize_risk(await _call_json("gemini-1.5-pro", prompt))


async def extract_events(body: str) -> list[dict] | None:
    bounded = _bounded_text(body)
    if not bounded:
        return None
    prompt = (
        "Extract real events or deadlines from this email body. "
        "Return JSON list only. Each item must have title, date, time, location, description, confidence. "
        "Only include items with confidence greater than 0.8.\n\n"
        f"BODY:\n{bounded}"
    )
    data = await _call_json("gemini-1.5-pro", prompt)
    if not isinstance(data, list):
        return None
    normalized: list[dict] = []
    for item in data:
        if not isinstance(item, dict):
            continue
        try:
            confidence = float(item.get("confidence", 0))
        except (TypeError, ValueError):
            confidence = 0
        if confidence <= 0.8:
            continue
        normalized.append(
            {
                "title": str(item.get("title", "")).strip()[:140] or "Event",
                "date": str(item.get("date", "")).strip() or None,
                "time": str(item.get("time", "")).strip() or None,
                "location": str(item.get("location", "")).strip()[:160] or None,
                "description": str(item.get("description", "")).strip()[:240] or None,
                "confidence": confidence,
            }
        )
    return normalized or None

File: backend2/src/services/gemini.ts
import { GoogleGenerativeAI } from "@google/generative-ai";

import { settings } from "../config.js";

let client: GoogleGenerativeAI | null = null;
let failureCount = 0;
let firstFailureTime = 0;
let circuitOpenUntil = 0;

const ALLOWED_PRIORITIES = new Set(["high", "medium", "low"]);
const ALLOWED_RISKS = new Set(["high", "medium", "low"]);
const ALLOWED_CATEGORIES = new Set(["primary", "updates", "promotions", "social", "forums"]);
const DEFAULT_FLASH_MODEL = "gemini-2.5-flash";
const FLASH_MODEL_FALLBACKS = [DEFAULT_FLASH_MODEL, "gemini-2.5-flash-lite", "gemini-2.0-flash"];

function ensureClient(): GoogleGenerativeAI {
  if (!client) client = new GoogleGenerativeAI(settings.geminiApiKey);
  return client;
}

function boundedText(value: string, limit = 6000): string {
  return value.replace(/\s+/g, " ").trim().slice(0, limit);
}

function circuitOpen(): boolean {
  return Date.now() < circuitOpenUntil;
}

function recordFailure(): void {
  const now = Date.now();
  if (!firstFailureTime || now - firstFailureTime > 60_000) {
    firstFailureTime = now;
    failureCount = 0;
  }
  failureCount += 1;
  if (failureCount >= 5) {
    circuitOpenUntil = now + 120_000;
    failureCount = 0;
    firstFailureTime = 0;
  }
}

function extractJson(text: string): any {
  const candidate = text.trim().replace(/^```(?:json)?/, "").replace(/```$/, "").trim();
  try {
    return JSON.parse(candidate);
  } catch {
    const startObj = candidate.indexOf("{");
    const endObj = candidate.lastIndexOf("}");
    const startList = candidate.indexOf("[");
    const endList = candidate.lastIndexOf("]");
    if (startObj !== -1 && endObj > startObj) return JSON.parse(candidate.slice(startObj, endObj + 1));
    if (startList !== -1 && endList > startList) return JSON.parse(candidate.slice(startList, endList + 1));
    throw new Error("Unable to parse Gemini JSON");
  }
}

async function callJson(modelName: string, prompt: string): Promise<any> {
  if (circuitOpen()) throw new Error("Gemini circuit open");
  try {
    const model = ensureClient().getGenerativeModel({
      model: modelName,
      generationConfig: { temperature: 0.15, responseMimeType: "application/json" }
    });
    const result = await model.generateContent(prompt);
    return extractJson(result.response.text());
  } catch (error) {
    recordFailure();
    throw error;
  }
}

function flashModel(): string {
  return settings.geminiModel ?? DEFAULT_FLASH_MODEL;
}

function flashModelCandidates(): string[] {
  const preferred = flashModel().trim();
  const ordered = [preferred, ...FLASH_MODEL_FALLBACKS];
  return [...new Set(ordered.filter(Boolean))];
}

async function callJsonWithFallbacks(modelNames: string[], prompt: string): Promise<any> {
  let lastError: unknown = null;
  for (const modelName of modelNames) {
    try {
      return await callJson(modelName, prompt);
    } catch (error) {
      lastError = error;
    }
  }
  throw lastError ?? new Error("Gemini request failed");
}

export async function summarizeEmail(text: string): Promise<string[] | null> {
  const bounded = boundedText(text);
  if (!bounded) return null;
  const data = await callJson(
    flashModel(),
    `Summarize this student-facing email into exactly 3 concise bullet points. Each bullet must be 6 to 15 words. Return JSON only as {"bullets":[..]}.\n\nEMAIL:\n${bounded}`
  );
  const bullets = Array.isArray(data?.bullets) ? data.bullets.map((item: unknown) => String(item).trim().slice(0, 120)).filter(Boolean) : [];
  return bullets.length === 3 ? bullets : null;
}

export async function priorityAnalysis(sender: string, subject: string, body: string): Promise<any | null> {
  const data = await callJson(
    flashModel(),
    `Assign priority for a student inbox. Consider deadlines, account actions, schedule changes, exams, finances, and urgency. Return JSON only as {priority:'high'|'medium'|'low',reason:string,confidence:float}.\n\nSENDER:${boundedText(sender, 200)}\nSUBJECT:${boundedText(subject, 300)}\nBODY:${boundedText(body)}`
  );
  const priority = String(data?.priority ?? "").toLowerCase();
  if (!ALLOWED_PRIORITIES.has(priority)) return null;
  return {
    priority,
    reason: String(data?.reason ?? "Priority inferred from message content.").slice(0, 240),
    confidence: Number(data?.confidence ?? 0.5)
  };
}

export async function categoryClassification(sender: string, subject: string, snippet: string): Promise<string | null> {
  const data = await callJson(
    flashModel(),
    `Classify a student email into one of: primary, updates, promotions, social, forums. Return JSON only as {category:'primary'|'updates'|'promotions'|'social'|'forums'}.\n\nSENDER:${boundedText(sender, 200)}\nSUBJECT:${boundedText(subject, 300)}\nSNIPPET:${boundedText(snippet, 600)}`
  );
  const category = String(data?.category ?? "").toLowerCase();
  return ALLOWED_CATEGORIES.has(category) ? category : null;
}

export async function phishingAnalysis(sender: string, subject: string, body: string): Promise<any | null> {
  const data = await callJson(
    "gemini-1.5-pro",
    `Analyze this email for phishing, impersonation, credential theft, payment fraud, or manipulative urgency. Return JSON only as {risk:'low'|'medium'|'high',reasons:[string]}.\n\nSENDER:${boundedText(sender, 200)}\nSUBJECT:${boundedText(subject, 300)}\nBODY:${boundedText(body)}`
  );
  const risk = String(data?.risk ?? "").toLowerCase();
  if (!ALLOWED_RISKS.has(risk)) return null;
  return {
    risk,
    reasons: Array.isArray(data?.reasons) ? data.reasons.map((item: unknown) => String(item).trim().slice(0, 160)).filter(Boolean).slice(0, 5) : []
  };
}

export async function extractEvents(body: string): Promise<any[] | null> {
  const bounded = boundedText(body);
  if (!bounded) return null;
  const data = await callJson(
    "gemini-1.5-pro",
    `Extract real events or deadlines from this email body. Return JSON list only. Each item must have title, date, time, location, description, confidence. Only include items with confidence greater than 0.8.\n\nBODY:\n${bounded}`
  );
  if (!Array.isArray(data)) return null;
  const normalized = data
    .filter((item) => Number(item?.confidence ?? 0) > 0.8)
    .map((item) => ({
      title: String(item.title ?? "Event").slice(0, 140),
      date: item.date ? String(item.date) : null,
      time: item.time ? String(item.time) : null,
      location: item.location ? String(item.location).slice(0, 160) : null,
      description: item.description ? String(item.description).slice(0, 240) : null,
      confidence: Number(item.confidence)
    }));
  return normalized.length ? normalized : null;
}

export async function emailInsights(
  sender: string,
  subject: string,
  body: string,
  question: string
): Promise<{
  answer: string;
  key_points: string[];
  suggested_action: string | null;
}> {
  const boundedBody = boundedText(body, 5000);
  const boundedQuestion = boundedText(question, 400);

  if (!boundedQuestion) {
    return {
      answer: "Please ask a question.",
      key_points: [],
      suggested_action: null
    };
  }

  const prompt =
    "You are an AI assistant helping a university student understand their email.\n" +
    "Given the email below, answer the user's question clearly and helpfully.\n" +
    'Return JSON only with this shape: {"answer":"string (2-4 sentences)","key_points":["string", ...],"suggested_action":"string or null"}\n\n' +
    `SENDER: ${boundedText(sender, 200)}\n` +
    `SUBJECT: ${boundedText(subject, 300)}\n` +
    `BODY:\n${boundedBody}\n\n` +
    `USER QUESTION: ${boundedQuestion}`;

  try {
    const data = await callJsonWithFallbacks(flashModelCandidates(), prompt);
    if (!data || typeof data !== "object" || Array.isArray(data)) {
      return {
        answer: "Could not parse AI response.",
        key_points: [],
        suggested_action: null
      };
    }

    const answer = String((data as any).answer ?? "").trim().slice(0, 800) || "No answer generated.";
    const keyPoints = Array.isArray((data as any).key_points)
      ? (data as any).key_points
          .map((point: unknown) => String(point).trim().slice(0, 200))
          .filter(Boolean)
          .slice(0, 5)
      : [];
    const suggestedAction = String((data as any).suggested_action ?? "").trim().slice(0, 300) || null;

    return {
      answer,
      key_points: keyPoints,
      suggested_action: suggestedAction
    };
  } catch {
    return {
      answer: "AI is temporarily unavailable. Please try again shortly.",
      key_points: [],
      suggested_action: null
    };
  }
}
